src/System.Linq/src/System/Linq/Concatenate.cs

```diff
@@ -111,26 +121,268 @@ internal override IEnumerable<TSource> GetEnumerable(int index)
+
+                ConcatIterator<TSource> previous = _previous;
+
+                // TODO: This can be made more readable once we use C# 7 which supports pattern matching.
+                ConcatNCollectionIterator<TSource> previousN;
+                while ((previousN = previous as ConcatNCollectionIterator<TSource>) != null)
+                {
+                    checked
+                    {
+                        copied += previousN._next.Count;
+                    }
+                    previousN._next.CopyTo(array, arrayIndex - copied);
+                    previous = previousN._previous;
+                }
+
+                // We've reached the first 2 collections that were concatenated
+                var previous2 = previous as Concat2CollectionIterator<TSource>;
+                Debug.Assert(previous2 != null);
+
+                copied += previous2.Count;
+                Debug.Assert(copied == Count); // We should have copied all the elements
+
+                previous2.CopyTo(array, arrayIndex - copied);
+            }
+
+            internal override IEnumerable<TSource> GetEnumerable(int index)
+            {
+                if (index > _nextIndex)
+                {
+                    return null;
+                }
+
+                ConcatNCollectionIterator<TSource> current = this;
                 while (true)
                 {
                     if (index == current._nextIndex)
                     {
                         return current._next;
                     }
 
-                    ConcatNIterator<TSource> prevN = current._previousConcat as ConcatNIterator<TSource>;
-                    if (prevN != null)
+                    var previousN = current._previous as ConcatNCollectionIterator<TSource>;
+                    if (previousN != null)
                     {
-                        current = prevN;
+                        current = previousN;
                         continue;
                     }
 
-                    Debug.Assert(current._previousConcat is Concat2Iterator<TSource>);
+                    var previous2 = current._previous as Concat2CollectionIterator<TSource>;
+                    Debug.Assert(previous2 != null);
                     Debug.Assert(index == 0 || index == 1);
-                    return current._previousConcat.GetEnumerable(index);
+                    return previous2.GetEnumerable(index);
                 }
             }
+
+            public override TSource[] ToArray()
+            {
+                int totalCount = Count;
+                if (totalCount == 0)
+                {
+                    return Array.Empty<TSource>();
+                }
+
+                var result = new TSource[totalCount];
+                CopyBefore(result, result.Length);
+                return result;
+            }
+
+            public override int GetCount(bool onlyIfCheap) => Count; // Getting the count is always cheap.
```

Comment:
@stephentoub I audited all usages of `GetCount(true)` from non-`GetCount` callers [here](https://gist.github.com/jamesqo/0f4623f5f485e624cb537c1ff8ea2619). It looks like all of the sites use it for pre-allocating an array/list of some sort, and if it's not cheap fall back to manually iterating and resizing the array/list. If the number of collections is large then the number of items is probably even larger, so this will undoubtedly lead to a lot of unnecessary resizing + increased memory consumption in cases like

```cs
largeSequenceOfStringCollections // IEnumerable<ICollection<string>>
    .Aggregate((accumulated, next) => accumulated.Concat(next)) // IEnumerable<string>
    .Append("foo") // IEnumerable<string>
    .ToArray(); // string[]
```